### PR TITLE
Gnosis safe

### PR DESCRIPTION
**Pull request by Bookcliff**
https://www.loom.com/share/55c6f52cc55646f592091313e11ce84c

- [x] Bond can be created with pending approval
- [x] Auction can be created with pending approval
- [x] Promissory note & auction can be created with pending bond creation

Issues #10  and #25  will be addressed in another PR

### CODE REVIEW

File: src/components/ProductCreate/BondActionSteps.tsx [hunk 1]
```diff
@@ -78,41 +84,61 @@ export const BondActionSteps = ({ convertible = true, disabled }) => {
...
+                      className={`pointer-events-auto hidden w-[85px] sm:btn`}
+                      // disabled={isGoerli}
+                      onClick={() => setScheduled(true)}
+                    >
+                      Approval Scheduled
+                    </button>
+                  </div>
+                </div>
+              }
+              tip="Bypass the approval process if the token approval has been scheduled in a previous transaction and is queued. IMPORTANT: Since the tokens have yet to be approved, gas will be automatically set to 1000000 Wei."
```
Comment (Namaskar-1F64F) [inline src/components/ProductCreate/BondActionSteps.tsx:138]: This surely needs to be in a separate component as it's used twice.

File: src/components/ProductCreate/selectors/CollateralTokenSelector.tsx [hunk 2]
```diff
@@ -62,8 +63,51 @@ const NoBondFound = () => {
...
   const { data, loading } = useBondsPortfolio()
+  // const { address } = useAccount()
+
+  const [n, setN] = useState(0)
+  const { provider } = useActiveWeb3React()
+
+  useEffect(() => {
+    async function GetNonce() {
+      const getNonce = await provider
+        .getTransactionCount('0xBE9A5b24dbEB65b21Fc91BD825257f5c4FE9c01D')
```
Comment (Namaskar-1F64F) [inline src/components/ProductCreate/selectors/CollateralTokenSelector.tsx:76]: use the bond factory constant

File: src/components/ProductCreate/selectors/CollateralTokenSelector.tsx [hunk 3]
```diff
@@ -62,8 +63,51 @@ const NoBondFound = () => {
...
+        .catch((e) => console.log({ e }))
+      const nonce = Number(getNonce)
+      const nonceSum = nonce
+      setN(nonceSum)
+    }
+    GetNonce()
+  }, [provider])
+
+  const from = '0xbe9a5b24dbeb65b21fc91bd825257f5c4fe9c01d'
```
Comment (Namaskar-1F64F) [inline src/components/ProductCreate/selectors/CollateralTokenSelector.tsx:85]: same thing with this. this doesn't work with mainnet either, right?

File: src/components/ProductCreate/selectors/CollateralTokenSelector.tsx [hunk 4]
```diff
@@ -62,8 +63,51 @@ const NoBondFound = () => {
...
+  const nonce = n
+
+  //WILL NEED TO CHANGE IF PAYMENT TOKEN ANYTHING OTHER THAN USDC
+
+  if (scheduled === true) {
```
Comment (Namaskar-1F64F) [inline src/components/ProductCreate/selectors/CollateralTokenSelector.tsx:90]:
```suggestion
  if (scheduled) {
```

File: src/components/ProductCreate/selectors/CollateralTokenSelector.tsx [hunk 5]
```diff
@@ -62,8 +63,51 @@ const NoBondFound = () => {
   )
 }
 
-export const BondSelector = () => {
+export const BondSelector = ({ scheduled = false }: { scheduled?: boolean }) => {
   const { data, loading } = useBondsPortfolio()
+  // const { address } = useAccount()
+
+  const [n, setN] = useState(0)
```
Comment (Namaskar-1F64F) [inline src/components/ProductCreate/selectors/CollateralTokenSelector.tsx:70]: I think this should be initialized to null and later checked if there is a value other than null. is 0 a valid nonce? if it is, the failure state may go through

File: src/components/ProductCreate/selectors/CollateralTokenSelector.tsx [hunk 6]
```diff
@@ -62,8 +63,51 @@ const NoBondFound = () => {
...
+
+  const from = '0xbe9a5b24dbeb65b21fc91bd825257f5c4fe9c01d'
+  const nonce = n
+
+  //WILL NEED TO CHANGE IF PAYMENT TOKEN ANYTHING OTHER THAN USDC
+
+  if (scheduled === true) {
+    const scheduledBond = window.localStorage.getItem('scheduledBond')
+    const scheduledBondArray = JSON.parse(scheduledBond)
+    const bondAddress = getContractAddress({ from, nonce })
```
Comment (Namaskar-1F64F) [inline src/components/ProductCreate/selectors/CollateralTokenSelector.tsx:93]: cool, so this is to get the future bond address when it's scheduled?

**Review by Namaskar-1F64F — CHANGES_REQUESTED** [review]
left some comments. looking good overall

File: src/components/CreditEnhancers/promissory/scheduledBondModal.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                </svg>
+                <span className="text-[#EDA651]">Warning</span>
+              </div>
+              <div className="text-sm text-[#9F9F9F]">
+                <p>
+                  The bond must already have been created with both the approval and creation
+                  transactions queued within the Gnosis Safe.
+                </p>
+                <p>
+                  Create a{' '}
```
[the remark this answers is not in the document]
Comment (Bookcliff) [reply src/components/CreditEnhancers/promissory/scheduledBondModal.tsx:35]: Fixed

File: src/components/OfferingCreate/auction/ActionSteps.tsx [hunk 8]
```diff
@@ -48,7 +51,10 @@ export const ActionSteps = ({ disabled }) => {
     ) {
       setCurrentApproveStep(1)
     }
-  }, [bondAllowance, auctionedSellAmount, bondToAuction.decimals])
+    if (scheduled) {
```
[the remark this answers is not in the document]
Comment (Bookcliff) [reply src/components/OfferingCreate/auction/ActionSteps.tsx:54]: Fixed

File: src/components/OfferingCreate/auction/StepOne.tsx [hunk 9]
```diff
@@ -1,26 +1,39 @@
...
 import TooltipElement from '@/components/common/Tooltip'
 import { currentTimeInUTC } from '@/utils/tools'
 
 export const StepOne = () => {
   const { register, watch } = useFormContext()
+  const walletType = window.localStorage.getItem('wagmi.wallet')
+  const [tip, setTip] = useState('')
+
+  useEffect(() => {
+    if (walletType === '"safe"') {
```
[the remark this answers is not in the document]
Comment (Bookcliff) [reply src/components/OfferingCreate/auction/StepOne.tsx:22]: Fixed

File: src/components/OfferingCreate/auction/StepOne.tsx [hunk 10]
```diff
@@ -1,26 +1,39 @@
...
+      setTip(
+        'Bond you will be selling. Select Scheduled Bond if you are scheduling an auction for a bond that has been queued.',
+      )
+    } else {
+      setTip('Bond you will be selling.')
+    }
+  }, [tip, walletType])
```
[the remark this answers is not in the document]
Comment (Bookcliff) [reply src/components/OfferingCreate/auction/StepOne.tsx:29]: Fixed

File: src/components/ProductCreate/selectors/CollateralTokenSelector.tsx (hunk 5 above)
Comment (Bookcliff) [reply src/components/ProductCreate/selectors/CollateralTokenSelector.tsx:70]: Fixed

File: src/components/ProductCreate/selectors/CollateralTokenSelector.tsx (hunk 2 above)
Comment (Bookcliff) [reply src/components/ProductCreate/selectors/CollateralTokenSelector.tsx:76]: Fixed

File: src/components/ProductCreate/selectors/CollateralTokenSelector.tsx (hunk 4 above)
Comment (Bookcliff) [reply src/components/ProductCreate/selectors/CollateralTokenSelector.tsx:90]: Fixed

File: src/components/ProductCreate/selectors/CollateralTokenSelector.tsx (hunk 6 above)
Comment (Bookcliff) [reply src/components/ProductCreate/selectors/CollateralTokenSelector.tsx:93]: Yep

File: src/components/CreditEnhancers/promissory/StepOne.tsx [hunk 11]
```diff
@@ -1,88 +1,129 @@
...
+  const [isScheduledBondOpen, setIsScheduledBondOpen] = useState(false)
+  return (
+    <>
+      <label className="label">
+        <>
+          <div className="label-text mr-2 flex w-full justify-between">
+            <span>
+              {labelName} <TooltipElement tip={tip} />
+            </span>
+            {walletType === '"safe"' && (
```
[the remark this answers is not in the document]
Comment (Bookcliff) [reply src/components/CreditEnhancers/promissory/StepOne.tsx:27]: No I think it is just if the wallet type is a gnosis safe. This variable is set through wagmi, not us

File: src/components/ProductCreate/BondActionSteps.tsx (hunk 1 above)
Comment (Bookcliff) [reply src/components/ProductCreate/BondActionSteps.tsx]: Fixed with your most recent merge

**Comment by Namaskar-1F64F** [comment]
this looks good, can you merge main into this and I think it's good to go. I added a fix for the regular wallet working with a scheduled transaction. It seemed to fail before. Might be worth testing the whole flow again. 

Also, should there be a way to clear the saved bond?

**Review by Namaskar-1F64F — APPROVED** [review]

**Comment by Bookcliff** [comment]
> Also, should there be a way to clear the saved bond?

I didn't think it was needed because if a user creates a new bond it will replace the saved bond. This also won't interfere with them selecting an already created bond, so I don't really see a reason for a user needing to clear the bond.

**Comment by Namaskar-1F64F** [comment]
thanks for this! looks great

**Pull request merged**